velox/functions/prestosql/ArrayDuplicates.cpp

```diff
@@ -0,0 +1,204 @@
+  void apply(
+      const SelectivityVector& rows,
+      std::vector<VectorPtr>& args, // Not using const ref so we can reuse args
+      const TypePtr& outputType,
+      exec::EvalCtx* context,
+      VectorPtr* result) const override {
+    // Acquire the array elements vector.
+    auto arrayVector = args.front()->as<ArrayVector>();
+    auto elementsVector = arrayVector->elements();
+    ArrayDuplicatesFunction::validateElementType(elementsVector->typeKind());
+
+    auto elementsRows =
+        toElementRows(elementsVector->size(), rows, arrayVector);
+    exec::LocalDecodedVector elements(context, *elementsVector, elementsRows);
+
+    vector_size_t numOfElements = elementsRows.size();
+    vector_size_t numOfRows = arrayVector->size();
+
+    // Allocate new vectors for indices, length and offsets.
+    memory::MemoryPool* pool = context->pool();
+    BufferPtr newIndices = allocateIndices(numOfElements, pool);
+    BufferPtr newSizes = allocateSizes(numOfRows, pool);
+    BufferPtr newOffsets = allocateOffsets(numOfRows, pool);
+
+    // Pointers and cursor to the raw data.
+    vector_size_t indexCursor = 0;
+    auto* rawIndices = newIndices->asMutable<vector_size_t>();
+    auto* rawSizes = newSizes->asMutable<vector_size_t>();
+    auto* rawOffsets = newOffsets->asMutable<vector_size_t>();
+
+    // Process the rows: use a hashmap to store unique values and
+    // whether it occurred once or more than once.
+    folly::F14FastMap<T, bool> uniqueMap;
+
+    rows.applyToSelected([&](vector_size_t row) {
+      auto size = arrayVector->sizeAt(row);
+      auto offset = arrayVector->offsetAt(row);
+
+      rawOffsets[row] = indexCursor;
+      vector_size_t numOfNulls = 0;
+
+      for (vector_size_t i = offset; i < offset + size; ++i) {
+        if (elements->isNullAt(i)) {
+          numOfNulls++;
+          if (numOfNulls == 2) {
+            rawIndices[indexCursor] = i;
+            indexCursor++;
+          }
+        } else {
+          T value = elements->valueAt<T>(i);
+          auto it = uniqueMap.find(value);
+          if (it == uniqueMap.end()) {
+            uniqueMap[value] = true;
+          } else if (it->second) {
+            it->second = false;
+            rawIndices[indexCursor] = i;
+            indexCursor++;
+          }
+        }
+      }
+
+      uniqueMap.clear();
+      rawSizes[row] = indexCursor - rawOffsets[row];
+
+      std::sort(
+          rawIndices + rawOffsets[row],
+          rawIndices + indexCursor,
+          lib::Index2ValueNullableLess<T>(elements));
+    });
+
+    newIndices->setSize(indexCursor * sizeof(vector_size_t));
+    auto newElements =
+        BaseVector::transpose(newIndices, std::move(elementsVector));
+
+    // Prepare and return result set.
+    auto resultArray = std::make_shared<ArrayVector>(
+        pool,
+        outputType,
+        nullptr,
+        numOfRows,
+        std::move(newOffsets),
+        std::move(newSizes),
+        std::move(newElements),
+        0);
```

Comment:
remove nulls count; this doesn't quite work and we will be revisiting metadata in vectors

Reply:
Thanks for providing the context!

I agree on overhauling the metadata including nullCount. I think it will be easier to do a one-off fix across all call sites, I'm not over-concerned about the correctness here along with other functions (e.g. array_distinct). Will leave it for now.